### ISSUE TITLE
MNIST distributed training and multi gpu tutorial

### ISSUE THREAD

**Issue by tarunjainsagar** [facebookarchive/caffe2#560]
How to convert MNIST python demo for 
    1. single and multiple GPU.
    2. distributed training ( multi-node , each node having multiple GPU) ??

**Comment by tarunjainsagar** [comment]
for single gpu - this works #374

**Comment by futurely** [comment]
https://caffe2.ai/docs/distributed-training.html

**Comment by tarunjainsagar** [comment]
Dear @futurely , I am aware of the caffe2 website. But, The code of resnet50_trainer.py is not that straight forward to match with our MNIST python demo tutorial. Please guide me towards something more specific.

Did you try MNIST on multiple GPU or Multi-Node ?

Although, this seems pretty useful, it gives a approach to work for multi-gpu 
https://caffe2.ai/docs/SynchronousSGD.html

**Comment by futurely** [comment]
resnet50_trainer.py is general enough to train any model on any dataset.

Change the dataset with the arguments `--train_data` & `--db_type`.
https://github.com/caffe2/caffe2/blob/master/caffe2/python/examples/resnet50_trainer.py#L313
```python
    # Input. Note that the reader must be shared with all GPUS.
    reader = train_model.CreateDB(
        "reader",
        db=args.train_data,
        db_type=args.db_type,
        num_shards=num_shards,
        shard_id=shard_id,
    )
```

ResNet-50 is hard coded but can be replaced with other networks.
https://github.com/caffe2/caffe2/blob/master/caffe2/python/examples/resnet50_trainer.py#L284
```python
    # Model building functions
    def create_resnet50_model_ops(model, loss_scale):
        [softmax, loss] = resnet.create_resnet50(
            model,
            "data",
            num_input_channels=args.num_channels,
            num_labels=args.num_labels,
            label="label",
            no_bias=True,
        )
        loss = model.Scale(loss, scale=loss_scale)
        model.Accuracy([softmax, "label"], "accuracy")
        return [loss]
```

**Comment by tarunjainsagar** [comment]
When I modify and replace the create_resnet50() with create_mnist()- After few iterations, I get Loss: NaN (not a number).
Code breaks at this point: https://github.com/caffe2/caffe2/blob/master/caffe2/python/examples/resnet50_trainer.py#L212

Create_mnist() contains the similar implementation as told in mnist python demo.

PS: I appreciate your quick responses. Thanks.

**Comment by tarunjainsagar** [comment]
okay, one quick question - Will putting the following in my create_mnist() function [ replacement of resenet.create_resnet50() ] solve my problem:
```

def create_mnist(
    model,
    data,
    num_input_channels,
    num_labels,
    label=None,
    is_test=False,
    no_loss=False,
    no_bias=0,
    conv1_kernel=7,
    conv1_stride=2,
    final_avg_kernel=7,
):
    # Image size: 28 x 28 -> 24 x 24
    conv1 = model.Conv(data, 'conv1', dim_in=1, dim_out=20, kernel=5)
    # Image size: 24 x 24 -> 12 x 12
    pool1 = model.MaxPool(conv1, 'pool1', kernel=2, stride=2)
    # Image size: 12 x 12 -> 8 x 8
    conv2 = model.Conv(pool1, 'conv2', dim_in=20, dim_out=50, kernel=5)
    # Image size: 8 x 8 -> 4 x 4
    pool2 = model.MaxPool(conv2, 'pool2', kernel=2, stride=2)
    # 50 * 4 * 4 stands for dim_out from previous layer multiplied by the image size
    fc3 = model.FC(pool2, 'fc3', dim_in=50 * 4 * 4, dim_out=500)
    fc3 = model.Relu(fc3, fc3)
    pred = model.FC(fc3, 'pred', 500, 10)

    (softmax, loss) = model.SoftmaxWithLoss(
            [pred, label],
            ["softmax", "loss"],
        )

   return (softmax, loss)
    
```

I get the following error on doing so ...
```
python mnist_cnn_single_multi_gpu_only.py --train_data ./mnist_lmdb/mnist-train-nchw-lmdb/ --epoch_size 60000 --num_epochs 2 --num_labels 10 --gpus 0 --num_channels 1 --db_type lmdb --file_store_path ./experiment_log
s/ 2>&1 | tee ./experiment_logs/try01.log
INFO:MNIST:Running on GPUs: [0]
INFO:MNIST:Using epoch size: 60000
INFO:data_parallel_model:Parallelizing model for devices: [0]
INFO:data_parallel_model:Create input and model training operators
INFO:data_parallel_model:Model for GPU: 0
INFO:data_parallel_model:Adding gradient operators
INFO:data_parallel_model:Add gradient all-reduces for SyncSGD
INFO:data_parallel_model:Post-iteration operators for updating params
INFO:data_parallel_model:Add initial parameter sync
WARNING:data_parallel_model:------- DEPRECATED API, please use data_parallel_model.OptimizeGradientMemory() -----
WARNING:memonger:NOTE: Executing memonger to optimize gradient memory
INFO:memonger:Remapping 9 blobs, using 2 shared
INFO:memonger:Memonger memory optimization took 0.000668048858643 secs
Could not create logging file: No such file or directory
COULD NOT CREATE A LOGGINGFILE 20170512-072417.20923!INFO:MNIST:Starting epoch 0/2
terminate called after throwing an instance of 'caffe2::EnforceNotMet'
  what():  [enforce fail at conv_op_cudnn.cc:235] filter.dim32(1) == C / group_. 1 vs 3 Error from operator:
input: "gpu_0/data" input: "gpu_0/conv1_w" input: "gpu_0/conv1_b" output: "gpu_0/conv1" name: "" type: "Conv" arg { name: "kernel" i: 5 } arg { name: "exhaustive_search" i: 1 } arg { name: "ws_nbytes_limit" i: 67108864 } arg { name: "order" s: "NCHW" } device_option { device_type: 1 cuda_gpu_id: 0 } engine: "CUDNN"
Ignoring @/caffe2/caffe2/contrib/nccl:nccl_ops as it is not a valid file.
Ignoring @/caffe2/caffe2/contrib/gloo:gloo_ops as it is not a valid file.
Ignoring @/caffe2/caffe2/contrib/gloo:gloo_ops_gpu as it is not a valid file.
*** Aborted at 1494588257 (unix time) try "date -d @1494588257" if you are using GNU date ***
PC: @     0x7fe64daa0c37 (unknown)
*** SIGABRT (@0x3ef000051bb) received by PID 20923 (TID 0x7ee60b93e700) from PID 20923; stack trace: ***
    @     0x7fe64de42330 (unknown)
    @     0x7fe64daa0c37 (unknown)
    @     0x7fe64daa4028 (unknown)
   0 bash

```

We need to provide DATA in MNIST DEMO TUTORIAL before creating model, here we do not do so in resnet50_trainer.py

**Comment by futurely** [comment]
https://github.com/BVLC/caffe/search?utf8=%E2%9C%93&q=loss+nan&type=Issues

**Comment by tarunjainsagar** [comment]
@futurely Please see above, i have updated my comment with error I get.

**Comment by futurely** [comment]
from caffe2.python.utils import DebugMode
DebugMode.run(main)

https://github.com/caffe2/caffe2/blob/200eae99fa2a08c2eff6d7c89d5a07d2468c5d27/caffe2/python/utils.py#L187

**Comment by tarunjainsagar** [comment]
@futurely using Debug Mode halts my system infinitely

**Comment by tarunjainsagar** [comment]
@Yangqing @bwasti @aaronmarkham  What could be possible reason for 
```
terminate called after throwing an instance of 'caffe2::EnforceNotMet'
  what():  [enforce fail at conv_op_cudnn.cc:235] filter.dim32(1) == C / group_. 1 vs 3 Error from operator:
```
command i ran:
`python mnist_cnn_single_multi_gpu_only_edited.py --train_data ./mnist_lmdb/mnist-train-nchw-lmdb/ --epoch_size 60000 --num_epochs 2 --num_labels 10 --gpus 0 --num_channels 1 --db_type lmdb --file_store_path ./experiment_logs/ 2>&1 | tee ./experiment_logs/try01.log`

**Comment by futurely** [comment]
The network probably expects `num_channels` to be 3.

**Comment by tarunjainsagar** [comment]
@futurely 
1.  No. It don't work with --num_channels 3, I already tried it.
2.  MNIST has grayscale images, so Number of Channels is 1.

**Comment by aaronmarkham** [comment]
You might want to take a look at this new tutorial. It covers using the GPU and is an adaptation of the resnet50_trainer script.
https://github.com/caffe2/caffe2/blob/master/caffe2/python/tutorials/Multi-GPU_Training.ipynb

**Comment by tarunjainsagar** [comment]
@futurely @aaronmarkham 
Example Implementation of mnist multi-node multi-gpu:
[multi_node_multi_gpu.txt](https://github.com/caffe2/caffe2/files/1028527/multi_node_multi_gpu.txt)

Command: `python mnist_cnn_single_multi_gpu_only_edited.py --train_data ./mnist_lmdb/mnist-train-nchw-lmdb/ --epoch_size 60000 --num_epochs 2 --num_labels 10 --gpus 0 --num_channels 1 --db_type lmdb --file_store_path ./experiment_logs/ 2>&1 | tee ./experiment_logs/try01.log
`

Please see the above implementation, which gives error 

```
COULD NOT CREATE A LOGGINGFILE 20170512-072417.20923!INFO:MNIST:Starting epoch 0/2
terminate called after throwing an instance of 'caffe2::EnforceNotMet'
  what():  [enforce fail at conv_op_cudnn.cc:235] filter.dim32(1) == C / group_. 1 vs 3 Error from operator:
input: "gpu_0/data" input: "gpu_0/conv1_w" input: "gpu_0/conv1_b" output: "gpu_0/conv1" name: "" type: "Conv" arg { name: "kernel" i: 5 } arg { name: "exhaustive_search" i: 1 } arg { name: "ws_nbytes_limit" i: 67108864 } arg { name: "order" s: "NCHW" } device_option { device_type: 1 cuda_gpu_id: 0 } engine: "CUDNN"
Ignoring @/caffe2/caffe2/contrib/nccl:nccl_ops as it is not a valid file.
```

**Comment by futurely** [comment]
cnn.CNNModelHelper has been replaced with model_helper.ModelHelper. 
https://github.com/caffe2/caffe2/blob/master/caffe2/python/tutorials/Multi-GPU_Training.ipynb

Do other datasets with either color or gray scale images work with the code? The problem may not lie in your code but in the MNIST dataset.

**Comment by futurely** [comment]
https://github.com/caffe2/caffe2/commit/3f4e52dac0ca9778f982174dedec00c69a57baac

**Comment by tarunjainsagar** [comment]
Dear @futurely , thanks for the continuous reponses.

problem exist even if I create my own lmdb data of grayscale images.

please ellaborate what you meant by 3f4e52d ; I could not understand it.

**Comment by futurely** [comment]
Since the MNIST tutorial can be run without using `data_parallel_model.Parallelize_GPU` and color image datasets can be trained with `data_parallel_model.Parallelize_GPU`, maybe there are indeed some bugs in `data_parallel_model` such as https://github.com/caffe2/caffe2/commit/3f4e52dac0ca9778f982174dedec00c69a57baac. Gray scale images are [not tested](https://github.com/caffe2/caffe2/blob/3f4e52dac0ca9778f982174dedec00c69a57baac/caffe2/python/data_parallel_model_test.py).

**Comment by aaronmarkham** [comment]
Maybe @pietern can comment. I know we tried this briefly even with 1 channel and met with similar issues.

**Comment by melody-rain** [comment]
I met the same issue. Can this be fixed?
